lib/drivers/bayerContourNext.js

```diff
@@ -271,110 +272,134 @@ module.exports = function (config) {
+      var record_type = struct.extractByte(record, 2);
+
+      if(record_type === 'R'){
+        parseDataRecord(record,function(err,r){
+          var robj = {};
+          if (err) {
+            debug('Failure trying to read record', record);
+            debug(err);
+            return callback(err, null);
+          } else {
+            robj.timestamp = parseInt(r[4]);
+            robj.annotations = getAnnotations(r[3], data);
+            robj.control = isControl(r[3]);
+            robj.units = r[2];
+            if(robj.units === 'mmol/L') {
+              robj.glucose = parseFloat(r[1]);
+            } else {
+              robj.glucose = parseInt(r[1]);
+            }
+            robj.nrec = parseInt(r[0]);
+            return callback(null, robj);
+          }
+        });
       } else {
-          robj.timestamp = parseInt(r[4]);
-          robj.annotations = getAnnotations(r[3], data);
-          robj.control = isControl(r[3]);
-          robj.units = r[2];
-          robj.glucose = parseInt(r[1]);
-          robj.nrec = parseInt(r[0]);
-          return callback(null, robj);
+        return callback(null, null);
       }
     });
   };
 
-  var bcnCommandResponse = function (commandpacket, callback) {
-
-    hidDevice.send(commandpacket.packet, function () {
-        receivePacket(5000, commandpacket.parser, function(err, packet) {
-            var rawtext = '';
-            while(hidDevice.hasAvailablePacket()){
-                var pkt = hidDevice.nextPacket();
-                pkt.parsed_payload = commandpacket.parser(pkt);
-                rawtext += pkt.parsed_payload;
-            }
+  var debugFrame = function (astmFrame) {
+    var output = astmFrame
+      .replace(/[\x02]/g, '<STX>')
+      .replace(/[\x03]/g, '<ETX>')
+      .replace(/[\x04]/g, '<EOT>')
+      .replace(/[\x05]/g, '<ENQ>')
+      .replace(/[\x06]/g, '<ACK>')
+      .replace(/[\x0A]/g, '<LF>')
+      .replace(/[\x0D]/g, '<CR>')
+      .replace(/[\x15]/g, '<NAK>')
+      .replace(/[\x17]/g, '<ETB>');
+    return output;
+  };
 
-            if (err !== 'TIMEOUT') {
-              callback(err, rawtext);
-            }else{
-              debug('receivePacket timed out');
-              callback(err, null);
-            }
-        });
+  var bcnCommandResponse = function (commandpacket, callback) {
+    hidDevice.send(commandpacket, function () {
+      getASTMFrame(5000, 3, function(err) {
+        if (err !== null || !hidDevice.hasAvailablePacket()) {
+          // TODO - we should do retries here, depending on the error
+          callback(err, null);
+        } else {
+          // We're using the hidDevice as a message/frame buffer, not a USB packet buffer
+          // to save on buffer copying. It could be refactored differently easily enough.
+          var frame = hidDevice.nextPacket();
+          // TODO - we should ideally check the validity of the ASTM frame (sequence number, checksum etc) here as well for retries
```

Comment:
I thought we are verifying the checksum, or is there a separate checksum for the ASTM frame itself?